r/tests/testthat/test-list-genes.R

```diff
@@ -0,0 +1,137 @@
+    pbmc_small@misc[["gene_dispersion"]] <- vars
+
+    return(pbmc_small)
+}
+
+test_that("List of genes generates the expected return format", {
+    data <- mock_scdata()
+    req <- mock_req()
+
+    res <- getList(req, data)
+
+    # number of genes is number of possible DE rows
+    expect_equal(res$full_count, nrow(data))
+
+    # returning only at most limit number of genes
+    expect_equal(nrow(res$gene_results), req$body$limit)
+
+    # ordering is correct
+    expect_equal(res$gene_results$dispersions, sort(res$gene_results$dispersions, decreasing = TRUE))
+
+    # have the correct column names
+    expect_columns <- c('gene_names', 'dispersions')
+    expect_equal(colnames(res$gene_results), expect_columns)
+})
+
+test_that("Order direction works properly", {
+    data <- mock_scdata()
+    req <- mock_req(orderBy = "dispersions", orderDirection = "DESC",offset = 0, limit = 20)
+
+    res <- getList(req, data)
+
+    # ordering is correct
+    expect_equal(res$gene_results$dispersions, sort(res$gene_results$dispersions, decreasing = TRUE))
+
+    req <- mock_req(orderBy = "dispersions", orderDirection = "ASC",offset = 0, limit = 20)
+
+    res <- getList(req, data)
+
+    # ordering is correct
+    expect_equal(res$gene_results$dispersions, sort(res$gene_results$dispersions, decreasing = FALSE))
+})
+
+test_that("Order by works properly", {
+    data <- mock_scdata()
+    req <- mock_req(orderBy = "gene_names", orderDirection = "DESC",offset = 0, limit = 20)
+
+    res <- getList(req, data)
+
+    # ordering is correct
+    expect_equal(res$gene_results$gene_names, sort(res$gene_results$gene_names, decreasing = TRUE))
+
+    req <- mock_req(orderBy = "gene_names", orderDirection = "ASC",offset = 0, limit = 20)
+
+    res <- getList(req, data)
+
+    # ordering is correct
+    expect_equal(res$gene_results$gene_names, sort(res$gene_results$gene_names, decreasing = FALSE))
+})
+
+test_that("Pagination works", {
+    data <- mock_scdata()
+    req <- mock_req(orderBy = "gene_names", orderDirection = "DESC",offset = 20, limit = 40)
+
+    res <- getList(req, data)
+
+    expect_equal(length(res$gene_results$gene_names), 40)
+
+    expect_equal(res$gene_results$gene_names,sort(rownames(data),decreasing=TRUE)[21:60])
+
+    req <- mock_req(orderBy = "dispersions", orderDirection = "DESC",offset = 20, limit = 40)
+
+    res <- getList(req, data)
+
+    expect_equal(length(res$gene_results$gene_names), 40)
+
+    expect_equal(res$gene_results$dispersions,sort(data@misc$gene_dispersion$variance.standardized,decreasing=TRUE)[21:60])
+})
+
+test_that("Start with pattern is applied", {
+    data <- mock_scdata()
+    req <- mock_req(orderBy = "gene_names", orderDirection = "DESC",offset = 0, limit = 40,geneNamesFilter = "^GZ")
+
+    res <- getList(req, data)
+
+    expect_true(all(grepl("^GZ",res$gene_results$gene_names)))
+    grep_results <- grepl("^GZ",rownames(data))
+    expect_equal(res$full_count,sum(grep_results==TRUE))
+})
```

Comment:
formatted, extracted pattern to variable. we can reuse this function changing pattern and test description.

```suggestion
test_that("Start with pattern is applied", {
    pat <- "^GZ"
    data <- mock_scdata()
    req <- mock_req(orderBy = "gene_names",
                    orderDirection = "DESC",
                    offset = 0,
                    limit = 40,
                    geneNamesFilter = pat)

    res <- getList(req, data)

    expect_true(all(grepl(pat, res$gene_results$gene_names)))
    
    grep_results <- grepl(pat, rownames(data))
    expect_setequal(res$gene_results$gene_names, grep_results)
    expect_equal(res$full_count, sum(grep_results == TRUE))
})
```